src/FuseGPUThreadLoops.cpp

```diff
@@ -372,6 +458,115 @@ class ExtractSharedAllocations : public IRMutator {
         }
     }
 
+    // Return index to free_spaces where 'alloc' should be coalesced. Return -1
+    // if there isn't any.
+    int find_best_fit(const vector<AllocPairs>& mem_allocs,
+                      const vector<int>& free_spaces,
+                      const SharedAllocation& alloc, int stage) {
+        int free_idx = -1;
+
+        Expr alloc_size = simplify(alloc.size);
+
+        // We prefer to coalesce dynamic-sized allocation with a dynamic-sized one and
+        // constant-sized alloc with a constant-sized one. If we can't find any free
+        // space with a matching type, we pick the most-recently freed space of the
+        // other type (e.g. pick constant-sized free space for a dynamic-sized allocation
+        // and vice versa). We prefer the most-recently freed space as stages that are
+        // close together usually have relatively similar allocation size. For
+        // constant-sized allocation, we prioritize free space which size differs
+        // the least with 'alloc' (can be smaller or larger; it does not really
+        // matter since we take the max of the two as the new size).
+
+        if (!is_const(alloc_size)) { // dynamic-sized alloc
+            for (int i = free_spaces.size() - 1; i >= 0; --i) {
+                // Only need to check the back of the vector since we always insert
+                // the most recent allocation at the back.
+                internal_assert(free_spaces[i] >= 0 && free_spaces[i] < (int)mem_allocs.size());
+                internal_assert(mem_allocs[free_spaces[i]].is_free(stage));
+
+                if (!is_const(mem_allocs[free_spaces[i]].max_size_bytes)) {
+                    return i;
+                } else if (free_idx == -1) {
+                    free_idx = i;
+                }
+            }
+        } else { // constant-sized alloc
+            int64_t diff = -1;
+            for (int i = free_spaces.size() - 1; i >= 0; --i) {
+                // Only need to check the back of the vector since we always insert
+                // the most recent allocation at the back.
+                internal_assert(free_spaces[i] >= 0 && free_spaces[i] < (int)mem_allocs.size());
+                internal_assert(mem_allocs[free_spaces[i]].is_free(stage));
+
+                if (is_const(mem_allocs[free_spaces[i]].max_size_bytes)) {
+                    Expr size = alloc_size * alloc.type.bytes();
+                    Expr dist = mem_allocs[free_spaces[i]].max_size_bytes - size;
+                    const int64_t *current_diff = as_const_int(simplify(dist * dist));
```

Comment:
abs(dist) makes more sense to me than dist*dist (avoids having to think about overflow)